src/cobra/io/sbml.py

```diff
@@ -1191,17 +1187,14 @@ def _model_to_sbml(cobra_model, f_replace=None, units=True):
         )
 
         # GPR
-        gpr = cobra_reaction.gene_reaction_rule
-        if gpr is not None and len(gpr) > 0:
-            # we will parse the GPR to format them correctly for libSBML
-            # avoids cases where a GPR can be parsed by cobrapy but not libSBML
-            tree, gids = parse_gpr(gpr)
+        gpr = cobra_reaction.gpr
+        if gpr is not None and hasattr(gpr, "body"):
             # replace ids in string
             if f_replace and F_GENE_REV in f_replace:
-                idmap = {gid: f_replace[F_GENE_REV](gid) for gid in gids}
-                gpr_new = ast2str(tree, names=idmap)
+                idmap = {gid: f_replace[F_GENE_REV](gid) for gid in gpr.geneset}
+                gpr_new = gpr.to_string(names=idmap)
             else:
-                gpr_new = ast2str(tree)
+                gpr_new = gpr.to_string()
```

Comment:
We should remember that this should be converted to SBML in a later PR.

Reply:
Basically we want to parse SBML directly by walking through it and translate to an AST and not go SBML -> str -> AST. The XML nodes in SBML can be mapped pretty much directly to the operations in an AST. In the original discussion I had a small example how to build the AST from scratch. This would speed up the parsing a lot, but is not required here.

I have no real preference where it lives (in GPR or in the SBML module). Since GPR also handles conversions to itself from other types (like str) you could add it as `GPR.from_sbml` or you could add a `sbml_to_grp` function to the SBML module in `io`. But however that gets implemented this should be a new PR.

Reply:
Okay.
Just to clarify a bit. SBML to AST is processed directly. GPR is converted directly form SBML if possible - I modified process_association(). See lines 705-714
Model to SBML uses r_fbc.createGeneProductAssociation() and setAssociation(), which are internal SBML functions, and I don't know how to replace them. I tried looking at SBML documentation and it was a bit complicated.